TESTING.md

```diff
@@ -0,0 +1,123 @@
+
+- [unit tests](#unit-tests)
+- [integration tests](#integration-tests)
+- [end to end (E2E) tests](#end-to-end-e2e-tests)
+
+### Unit tests
+
+Unit tests verify the functionality of one or more `func` or `struct`s
+in an isolated environment.
+
+These tests are independent of the environment they are run in and require no
+setup or teardown code.
+
+> **NOTE**: Some of these tests do require [`envtest`][envtest] setup to be present e.g.
+> [this one][test_GetAdminAPIsForService] or [this one][test_HTTPRouteReconcilerProperlyReactsToReferenceGrant].
+> This setup is handled automatically via `setup-envtest` Makefile target which is a dependency
+> of mentioned above unit test targets.
+>
+> As part of [#4099][issue4099] this organization might change and tests that do
+> require this setup might end up in a separate place.
+
+These tests can either be written in a way that do not require any mocks or fakes,
+or using e.g. kubernetes related fakes like for instance
+[`TestAddressesFromEndpointSlice`][test_AddressesFromEndpointSlice] does.
+
+#### How to run
+
+Unit tests can be either run via appropriate `go test ...` invocation or using any of
+the provided Makefile targets:
+
+- `test.unit` run unit tests with standard verbose output on stderr
+- `test.unit.pretty` run unit tests with opinionated output format, one line per package
+
+Unit tests are typically located in the same package as the code they test
+and they do not use any [build tags][go_build_tag].
+
+> **NOTE**: Some of these tests that do require [`envtest`][envtest] setup and
+> use a special [build tags][go_build_tag]: `envtest`. This is in place so that
+> simply running `go test ...` without any special build tags would pass without
+> requiring any manual setup from the developer.
+
+[test_GetAdminAPIsForService]: https://github.com/Kong/kubernetes-ingress-controller/blob/753e91f73dea5e51a3610d50c8a5928da79baa0f/internal/adminapi/endpoints_envtest_test.go#L28
+[test_HTTPRouteReconcilerProperlyReactsToReferenceGrant]: https://github.com/Kong/kubernetes-ingress-controller/blob/ccafa7ca9da7fb52ba959c2ebbc0974e22497b5b/internal/controllers/gateway/httproute_controller_envtest_test.go#L37
+[test_AddressesFromEndpointSlice]: https://github.com/Kong/kubernetes-ingress-controller/blob/753e91f73dea5e51a3610d50c8a5928da79baa0f/internal/adminapi/endpoints_test.go#L23-L24
+[issue4099]: https://github.com/Kong/kubernetes-ingress-controller/issues/4099
+[envtest]: https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/envtest
+
+### Integration tests
+
+KIC's integration tests rely on its controller manager being run in the same
+process as the tests via [`manager.Run()`][manager_run].
+
+These tests rely on a [test suite setup][integration_test_suite]
+using [`TestMain()`][pkggodev_testmain].
+
+Said setup will either create a new cluster or use an existing one to run the
+tests against.
+
+Most of the setup - and cleanup after the tests have run - is being done using
```

Comment:
Maybe move it to the section `## Testing levels and say that KTF is used for both integration and e2e for set up and tear down. Wdyt?

Reply:
This is not 100% the case because in some places we also have some custom cleanup logic.

I'd maybe not mention this way - bundling integration and e2e tests together. WDYT?

Reply:
yup, good point